# Тестовое задание
на  позицию «Data analyst / Product Analyst »



Компания: iGaming Company     
Исполнитель: Сасимович Игорь (ihar.sasimovich@gmail.com, Telegram: http://t.me/ihar367)

# Задание 1 (теория, KPI)

Составьте список KPI метрик необходимых для понимания качества
интернет-казино. Ответ аргументируйте.

### Финансовые KPI

* <span style="color: #FF4500;">GGR (Gross Gaming Revenue)</span> — валовая прибыль казино.  
Суть: показывает разницу между общей суммой денежных поступлений от игроков и суммой выигранных ими средств. Этот показатель демонстрирует доход компании до вычета налогов, операционных расходов и других затрат.

* <span style="color: #FF4500;">NGR (Net Gaming Revenue)</span> — валовая прибыль казино (GGR) за вычетом бонусов, возвратов, комиссий операторам платежных систем, роялти для разработчиков игр, лицензионных выплат и налогов.  
Суть: отражает чистую игровую прибыль, давая полное представление об эффективности проекта и реальной прибыли компании от игрового продукта после учета всех необходимых расходов.

* <span style="color: #FF4500;">NGR-to-Deposits</span> — показывает процент пользовательских депозитов, который превращается в чистую прибыль казино.   
Суть: позволяет понять, какая доля депозитов игроков фактически преобразуется в реальную прибыль компании и сколько чистой прибыли приходится на каждый вложенный игроком доллар.

* <span style="color: #FF4500;">Bets-to-Deposits</span> — показывает, какую часть своих депозитов игроки используют для ставок.   
Суть: просто депозиты сами по себе не приносят пользы для баланса казино. Для создания ценности важно, чтобы игроки проявляли желание рисковать и делали ставки. Этот показатель отражает активность игроков и их готовность использовать внесённые средства для ставок.

* <span style="color: #FF4500;">Средний депозит на игрока (Average Deposit Per Player)</span> — метрика, показывающая среднюю сумму денег, внесённую одним игроком за определённый период.   
Суть: позволяет оценить уровень платёжеспособности игроков и выявить изменения в депозитной активности.

* <span style="color: #FF4500;">Процент выплат (Payout Percentage)</span> — показатель, отражающий долю от общей суммы ставок, которая возвращается игрокам в виде выигрышей.   
Суть: метрика, которая балансирует между удержанием игроков и финансовой стабильностью казино. Высокий процент выплат способствует повышению лояльности игроков, но может снижать прибыль компании. Низкий процент, наоборот, увеличивает доход казино, но может негативно сказаться на удержании игроков.

* <span style="color: #FF4500;">ARPU (Average Revenue Per User)</span> — это средний объем прибыли, которую приносит казино один игрок за определенный период времени, как правило месяц.    
Суть: метрика, которая позволяет оценить общий доход, получаемый от всех пользователей, включая как платящих, так и неплатящих. Это помогает оценить общую эффективность продукта или услуги.



### Метрики удержания и активности

* <span style="color: #FF4500;">Retention Rate (RR)</span> — показатель удержания пользователей, который помогает оценить, сколько игроков продолжают активно пользоваться сервисом в определённый период времени после регистрации или первого визита.   
Суть: показывает, сколько людей вернулись (процент пользователей) в день Х (неделю Х, месяц Х) после первого входа в продукт (Day 0). Эта метрика помогает понять, насколько успешно казино удерживает пользователей и насколько эффективны стратегии вовлечения и удержания клиентов.

* <span style="color: #FF4500;">Churn Rate</span> — показатель оттока пользователей. Метрика обратная Retention.   
Суть: измеряет процент пользователей, которые перестали использовать сервис или покинули продукт за определённый период времени. Этот показатель важен для оценки стабильности пользовательской базы и помогает определить, насколько эффективно компания удерживает своих клиентов.

* <span style="color: #FF4500;">DAU, WAU, MAU (Daily/Weekly/Monthly Active Users)</span> — метрики, которые показывают уровень активности пользователей в разные временные промежутки.   
Суть: измеряют количество уникальных пользователей, взаимодействующих с продуктом в течение соответствующего периода (дня, недели, месяца).

* <span style="color: #FF4500;">Среднее время игровой сессии (Average Session Duration)</span> — метрика, показывающая среднее количество времени, которое игроки проводят в игре за одну сессию.   
Суть: отражает вовлечённость пользователей и помогает оценить, насколько долго игроки остаются в игре. Чем выше среднее время сессии, тем более увлекательным и захватывающим воспринимается продукт.

* <span style="color: #FF4500;">Частота визитов (Session Frequency)</span> — метрика, оценивающая, как часто игроки заходят в игру за определённый период времени.   
Суть: показывает регулярность визитов пользователей, что помогает понять, насколько привлекателен и удерживающий продукт. Высокая частота визитов может свидетельствовать о хорошем вовлечении игроков и сильной заинтересованности в игре.


### Метрики конверсии

* <span style="color: #FF4500;">Conversion Rate (в регистрацию, в депозит, во второй депозит и последующие)</span> — метрика, показывающая эффективность маркетинга и пользовательского пути.   
Суть: измеряет процент пользователей, которые переходят через ключевые этапы: от регистрации к первому депозиту, от первого депозита ко второму и от второго депозита к повторным платежам. Этот показатель помогает понять, насколько успешны стратегии привлечения и удержания пользователей, а также оценить эффективность конверсии на каждом из этапов взаимодействия.


### Маркетинговые метрики

* <span style="color: #FF4500;">LTV (Lifetime Value),  она же Customer Lifetime Value (CLV)</span> – это показатель ценности клиента, которую он приносит за все время в проекте. Он показывает, сколько денег в среднем принесет один пользователь за все время использования продукта.
Суть: позволяет оценить общую ценность пользователя для проекта, что помогает определить, сколько можно потратить на его привлечение. В идеале, LTV должен превышать CPA (Cost Per Acquisition) — стоимость привлечения пользователя, что гарантирует прибыльность маркетинговых усилий.

* <span style="color: #FF4500;">CPA (Cost Per Acquisition)</span> — это метрика, которая определяет стоимость привлечения одного нового пользователя в проект.
Суть: помогает оценить, сколько денег требуется для привлечения одного пользователя через маркетинговые кампании или другие каналы. Эффективность маркетинга обеспечивается, когда CPA меньше или равен LTV (Lifetime Value), что гарантирует прибыльность вложений в привлечение пользователей.

* <span style="color: #FF4500;">ROI (Return on Investment)</span> — метрика, показывающая прибыль или убыток от инвестиций в проект по сравнению с затратами.   
Суть: позволяет оценить эффективность вложенных средств, показывая, сколько прибыли (или убытков) приносит каждый вложенный доллар.


### Метрики пользовательского опыта и удовлетворенности

* <span style="color: #FF4500;">Net Promoter Score (NPS)</span> — метрика, показывающая, готовы ли игроки рекомендовать казино другим.   
Суть: оценивает лояльность пользователей, измеряя, сколько из них готовы порекомендовать казино своим друзьям или коллегам. Высокий NPS свидетельствует о сильной клиентской лояльности и хорошем репутационном имидже бренда.

* <span style="color: #FF4500;">Количество жалоб и среднее время их обработки</span> — важные показатели качества клиентской поддержки.   
Суть: количество жалоб помогает оценить общий уровень недовольства игроков, а среднее время их обработки — эффективность работы службы поддержки. Эти показатели влияют на удержание клиентов и их удовлетворенность сервисом.

* <span style="color: #FF4500;">Доля положительных отзывов о казино и поддержке</span> — качественная метрика для понимания репутации казино.   
Суть: измеряет процент положительных отзывов пользователей о казино и обслуживании. Высокая доля положительных отзывов подтверждает высокое качество услуг и способствует укреплению репутации бренда.

-----

KPI взаимосвязаны между собой и постоянный контроль за KPI является необходимым условием для того, чтобы онлайн-казино оставалось прибыльным и устойчивым в условиях конкуренции на долгие годы.
Используя комплексный подход, включающий финансовые, маркетинговые и операционные метрики, можно получить полное представление о качестве работы интернет-казино. Эти KPI помогают не только оценить доходность, но и анализировать показатели удержания пользователей, их вовлечённость и удовлетворённость. Полученные данные позволяют оптимизировать работу казино и разрабатывать стратегии, направленные на улучшение пользовательского опыта, повышение лояльности клиентов и <span style="color: #FF4500;">максимизацию прибыли</span>.

# Задание 2 (SQL, ClickHouse)

У вас дана SQL-база (ClickHouse), содержащая следующие таблицы:

* event_log- таблица содержит информацию о событиях игроков:

| Наименование поля | Тип данных      | Описание                       |
|-------------------|-----------------|---------------------------------|
| `event_timestamp` | `DateTime64`    | Время события                  |
| `player_id`       | `Int64`         | Идентификатор пользователя     |
| `play_session_id` | `Int64`         | Идентификатор сессии           |
| `gadget_id`       | `Int64`         | Идентификатор устройства       |
| `gadget_kind`     | `String`        | Тип устройства                 |
| `os_name`         | `String`        | Операционная система устройства|
| `country_code`    | `String`        | Код страны (ISO)               |
| `activity_name`   | `String`        | Название события               |

* user_info - таблица содержит информацию о пользователях:

| Наименование поля        | Тип данных    | Описание                                                 |
|---------------------------|---------------|-----------------------------------------------------------|
| `player_id`              | `Int64`       | ID пользователя на проекте                               |
| `contact_email`          | `String`      | Электронная почта пользователя                           |
| `record_created`         | `DateTime64`  | Дата регистрации                                         |
| `verification_email_sent`| `DateTime64`  | Дата отправки письма с подтверждением почты              |
|                           |               | Если `NULL`, то значение по умолчанию: `1970-01-01`      |
| `email_verified`         | `DateTime64`  | Дата подтверждения почты пользователем                   |
|                           |               | Если `NULL`, то значение по умолчанию: `1970-01-01`      |
| `country_code`           | `String`      | Код страны (ISO)                                         |


* payment_records - таблица платежей пользователей:

| Наименование поля     | Тип данных   | Описание                                                    |
|-----------------------|--------------|-------------------------------------------------------------|
| `transaction_id`      | `Int64`      | Идентификатор транзакции                                    |
| `player_id`           | `Int64`      | Идентификатор пользователя                                  |
| `payment_successful`  | `Int8`       | Статус успеха платежа (`1` — успешный, `0` — неуспешный)    |
| `payment_initiated`   | `DateTime64` | Дата создания платежа                                       |
| `payment_completed`   | `DateTime64` | Дата проведения платежа. Если `NULL`, то значение: `1970-01-01` |
| `payment_currency`    | `String`     | Валюта платежа                                              |
| `amount`              | `Float32`    | Размер платежа                                              |
| `amount_RUB`          | `Float32`    | Размер платежа в RUB                                        |


### На основании имеющихся таблиц необходимо написать SQL-запросы:


In [1]:
import clickhouse_connect
import pandas as pd

In [2]:
# Подключимся к ранее созданной БД db_maxbitsolution с помощью библиотеки clickhouse-connect

client = clickhouse_connect.get_client(
    host='localhost',
    port=8123,
    username='default',
    password='maz643028',
    database='db_maxbitsolution'
)

Не успел сгенерировать данные, залил по 20 строк 🥲

In [3]:
# Создадим функцию для обработки запросов
def select(sql: str) -> pd.DataFrame:
    result = client.query(sql)
    return pd.DataFrame(result.result_rows, columns=result.column_names)

1) который посчитает количество уникальных устройств и уникальных пользователей по месяцам в период с Июля-2023 по Декабрь-2023 (включительно);

In [4]:
query_1 = """
SELECT
  toYYYYMM(event_timestamp) AS month,
  COUNT(DISTINCT gadget_id) AS unique_devices,
  COUNT(DISTINCT player_id) AS unique_users
FROM event_log
WHERE event_timestamp BETWEEN '2023-07-01' AND '2023-12-31'
GROUP BY month
ORDER BY month;
""".strip()

select(query_1)

month  unique_devices  unique_users
0  202307               3             3
1  202308               3             3
2  202309               3             3
3  202310               3             3
4  202311               3             3
5  202312               4             4

2) который выведет ТОП-5 стран по количеству зарегистрированных пользователей в 2023 году;

In [5]:
query_2 = """
SELECT
  country_code,
  COUNT(player_id) AS user_count
FROM user_info
WHERE toYear(record_created) = '2023'
GROUP BY country_code
ORDER BY user_count DESC
LIMIT 5;
""".strip()

select(query_2)

country_code  user_count
0           DE           3
1           JP           3
2           RU           3
3           US           3
4           FR           3

3) который выведет ТОП-5 пользователей по общей сумме платежей, и по каждой стране за 2023 год;

In [6]:
query_3 = """
WITH ranked_users AS (
  SELECT
    country_code,
    player_id,
    SUM(amount_RUB) AS total_payment,
    ROW_NUMBER() OVER (PARTITION BY country_code ORDER BY SUM(amount_RUB) DESC) AS rank
  FROM payment_records
  JOIN user_info USING (player_id)
  WHERE payment_successful = 1 AND toYear(payment_completed) = '2023'
  GROUP BY country_code, player_id
)
SELECT country_code, player_id, total_payment
FROM ranked_users
WHERE rank <= 5;
""".strip()

select(query_3)

country_code  player_id  total_payment
0            IT         13        14000.0
1            ES         15         9600.0
2            ES          8         6000.0
3            JP          3         7000.0
4            JP         10         5600.0
5            RU          5         5000.0
6            DE         11        12000.0
7            DE          4         6000.0
8            US          7         8000.0
9            US         14         7600.0
10           US          1         4000.0

4) который посчитает среднюю продолжительности сессии на сайте за 2023 год;

In [7]:
query_4 = """
SELECT 
   AVG(session_duration) AS avg_session_duration
FROM (
  SELECT 
    play_session_id, 
    MAX(event_timestamp) - MIN(event_timestamp) AS session_duration
  FROM event_log
  WHERE toYear(event_timestamp) = 2023
  GROUP BY play_session_id
) AS sessions;
""".strip()

select(query_4)

avg_session_duration
0                   0.0

5) который посчитает количество уникальных пользователей, посетивших сайт (DAU) по месяцам, которые подтвердили свой email-адрес. Аналитика нужна за период с Июля-2023 по Декабрь-2023 (включительно);

In [8]:
query_5 = """
SELECT
  toYYYYMM(event_timestamp) AS month,
  COUNT(DISTINCT player_id) AS MAU
FROM event_log
JOIN user_info USING (player_id)
WHERE event_timestamp BETWEEN '2023-07-01' AND '2023-12-31'
  AND email_verified != '1970-01-01'
GROUP BY month
ORDER BY month;
""".strip()

select(query_5)

# Как я понял, в задании имелось ввиду не DAU, а MAU.

month  MAU
0  202307    3
1  202308    3
2  202309    3
3  202310    3
4  202311    3
5  202312    4

6) который позволит оценить долю (по сумме успешного депозита) пользователей с подтвержденной почтой, в Декабре 2023 года;

In [9]:
query_6 = """
SELECT
  SUM(CASE WHEN email_verified != '1970-01-01' THEN amount_RUB ELSE 0 END) / SUM(amount_RUB) AS verified_user_share
FROM payment_records
JOIN user_info USING (player_id)
WHERE payment_successful = 1
  AND payment_completed >= '2023-12-01' 
  AND payment_completed < '2024-01-01'
""".strip()

select(query_6)

verified_user_share
0                  NaN

7) который выведет для каждого пользователя среднее значение, нижний квартиль, медиану и верхний квартиль по времени между успешными депозитами.

In [10]:
query_7 = """
SELECT
  player_id,
  AVG(time_between_deposits) / 3600 AS avg_time_between_deposits_hours,
  quantile(0.25)(time_between_deposits) / 3600 AS lower_quartile_hours,
  quantile(0.5)(time_between_deposits) / 3600 AS median_time_hours,
  quantile(0.75)(time_between_deposits) / 3600 AS upper_quartile_hours
FROM (
  SELECT
    player_id,
    dateDiff('second', 
             lagInFrame(payment_completed) OVER (PARTITION BY player_id ORDER BY payment_completed),
             payment_completed) AS time_between_deposits
  FROM payment_records
  WHERE payment_successful = 1
    AND payment_completed != '1970-01-01 00:00:00'
) AS time_data
WHERE time_between_deposits > 0
GROUP BY player_id
HAVING COUNT(time_between_deposits) > 1; 
""".strip()

select(query_7)

# HAVING COUNT(time_between_deposits) > 1 --- д.б. минимум 2 успешных депозита

Empty DataFrame
Columns: []
Index: []

# Задание 3 (манипуляции данными в Python)

<br>
Есть таблица с действиями пользователей A/B-теста (<i>task_3_events.parquet</i>):

$events$ - посещение пользователем сайта
- $user\_id$ - id юзера,
- $user\_group$ - группа A/B-теста,
- $time$ - время совершения действия,

Пользовательская сессия определяется по следующим правилам:
<br>
1. Новая сессия начинается после 30 минут бездействия.

<b>Постройте таблицу с сессиями юзеров в формате<b>:
<br>
- $user\_id$ - id юзера
- $ab\_group$ - группа A/B-теста,
- $session\_start$ - время старта сессии,
- $session\_end$ - время окончания сессии,
- $session\_date$ - дата сессии.

In [11]:
df = pd.read_parquet(r'D:\000_Jupyter_notebooks\2025 02 Maxbitsolution, Data-analyst\task_3_events.parquet')
df.head()

user_id  user_group                       time
0  test_user_68722185           0 2023-12-01 00:00:02.465889
1  test_user_50279722           1 2023-12-01 00:00:02.490361
2  test_user_68722185           0 2023-12-01 00:00:03.803897
3  test_user_68721973           0 2023-12-01 00:00:03.821829
4  test_user_68721973           0 2023-12-01 00:00:03.854164

In [12]:
df.dtypes

user_id               object
user_group             int64
time          datetime64[ns]
dtype: object

In [13]:
df.isna().sum()

user_id       0
user_group    0
time          0
dtype: int64

In [14]:
df = df.sort_values(by=['user_id', 'time'])

In [15]:
# Сдвигаем временные значения для расчета разрывов между действиями
df['prev_time'] = df.groupby('user_id')['time'].shift()

In [16]:
# Рассчитываем разрывы между действиями (в минутах)
df['time_diff'] = (df['time'] - df['prev_time']).dt.total_seconds() / 60

In [17]:
# Определяем начало новой сессии (разрыв > 30 минут или первое событие)
df['new_session'] = (df['time_diff'] > 30) | (df['prev_time'].isna())

In [18]:
# Генерируем метку для каждой сессии
df['session_id'] = df.groupby('user_id')['new_session'].cumsum()

In [19]:
# Группируем для формирования сессий
session_df = df.groupby(['user_id', 'user_group', 'session_id']).agg(
    session_start=('time', 'min'),
    session_end=('time', 'max')
).reset_index()

In [20]:
session_df['session_date'] = session_df['session_start'].dt.date

In [21]:
session_df = session_df.rename(columns={'user_group': 'ab_group'})

In [22]:
session_df = session_df.drop(columns=['session_id'])

In [23]:
session_df.head()

user_id  ab_group              session_start  \
0  test_user_10021823         1 2023-12-07 01:08:51.008309   
1  test_user_10030144         0 2023-12-02 20:37:34.861127   
2   test_user_1003396         1 2023-12-01 00:00:48.238407   
3   test_user_1003396         1 2023-12-02 08:21:13.134305   
4   test_user_1003396         1 2023-12-02 21:12:09.512252   

                 session_end session_date  
0 2023-12-07 01:10:11.804765   2023-12-07  
1 2023-12-02 20:38:14.301560   2023-12-02  
2 2023-12-01 00:01:41.376842   2023-12-01  
3 2023-12-02 08:35:03.351138   2023-12-02  
4 2023-12-02 21:34:53.437896   2023-12-02

In [24]:
session_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25044 entries, 0 to 25043
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   user_id        25044 non-null  object        
 1   ab_group       25044 non-null  int64         
 2   session_start  25044 non-null  datetime64[ns]
 3   session_end    25044 non-null  datetime64[ns]
 4   session_date   25044 non-null  object        
dtypes: datetime64[ns](2), int64(1), object(2)
memory usage: 978.4+ KB


# Задание 3 (оценка АБ-теста)

Был проведен A/B-тест.
<br>
В качестве данных используйте таблицу, построенную в предыдущем задании. Первая сессия юзера считается моментом попадания в A/B-тест.
<br><br>
Ключевая метрика эксперимента - конверсия во вторую сессию.
<br>
Сделайте вывод о том, какая группа выиграла в A/B-тесте. Ответ обоснуйте.
<br>

In [25]:
# Проверим на наличие дублирующих строк 
session_df.duplicated().sum()

0

In [26]:
# Проверим на наличие пропущенных значений
session_df.isna().sum()

user_id          0
ab_group         0
session_start    0
session_end      0
session_date     0
dtype: int64

In [27]:
session_counts = session_df.groupby(['user_id', 'ab_group'])['session_date'].nunique().reset_index()
session_counts.columns = ['user_id', 'ab_group', 'session_count']
session_counts.head()

user_id  ab_group  session_count
0  test_user_10021823         1              1
1  test_user_10030144         0              1
2   test_user_1003396         1              8
3  test_user_10035974         0              1
4   test_user_1005039         1              1

In [28]:
session_counts['converted'] = session_counts['session_count'] > 1
session_counts.head()

user_id  ab_group  session_count  converted
0  test_user_10021823         1              1      False
1  test_user_10030144         0              1      False
2   test_user_1003396         1              8       True
3  test_user_10035974         0              1      False
4   test_user_1005039         1              1      False

In [29]:
conversion_rates = session_counts.groupby('ab_group')['converted'].mean().reset_index()
conversion_rates.columns = ['ab_group', 'conversion_rate']
conversion_rates

# Контрольная группа А: ab_group = 0
# Тестовая группа B: ab_group = 1

ab_group  conversion_rate
0         0         0.194373
1         1         0.190026

Результат: Конверсия во вторую сессию составляет 19.44% для контрольной группы и 19.00% для тестовой. Случайна ли разница, не очевидно. Требуется дополнительный статистический анализ для определения значимости разницы.

Примененим тест хи-квадрат.

Определимся с гиппотезами:

* H0: Разница между коэффициентами конверсии тестовой и контрольной групп статистически не значима. Выборки принадлежат одной генеральной совокупности, различий нет, нововведения не оказывают влияния.

* H1: Разница между коэффициентами конверсии тестовой и контрольной групп статистически значима. Различия есть.

In [30]:
from scipy.stats import chi2_contingency


In [31]:
conversion_data = session_counts.groupby('ab_group')['converted'].value_counts().unstack().fillna(0)
conversion_data 

converted  False  True 
ab_group               
0           5985   1444
1           5993   1406

In [32]:
# Применение теста хи-квадрат
chi2_stat, p_value, dof, expected = chi2_contingency(conversion_data)

print(f"\nХи-квадрат статистика: {chi2_stat:.4f}")
print(f"P-значение: {p_value:.4f}")

if p_value < 0.05:
    print("Разница статистически значима.")
else:
    print("Разница не является статистически значимой.")


Хи-квадрат статистика: 0.4237
P-значение: 0.5151
Разница не является статистически значимой.


На имеющихся данных, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативной. P-значение хи-квадрат теста составило 0.5151, что значительно превышает уровень значимости 0.05. Это означает, что наблюдаемая разница между коэффициентами конверсии тестовой (19.00%) и контрольной (19.44%) групп может быть объяснена случайными колебаниями, а не реальным эффектом нововведений.

При таком исходе бизнесу стоит рассмотреть следующие шаги:

* Проверить другие гипотезы, направленные на улучшение пользовательского поведения.
* Провести тестирование с более чувствительными метриками, такими как время между сессиями, вовлеченность или средний чек (ARPU).
* Анализировать данные по сегментам пользователей, чтобы выявить возможные локальные эффекты.
* Рассмотреть возможность увеличения выборки для повышения статистической мощности теста.